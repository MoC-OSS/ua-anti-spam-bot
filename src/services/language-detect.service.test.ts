import { languageDetectUaMocks } from './_mocks/language-detect.mocks';
import { languageDetectService } from './language-detect.service';

describe('LanguageDetectService', () => {
  describe('isRussian', () => {
    it('should return true if russian', () => {
      const text = '—Å—ä–µ—à—å –µ—â–µ —ç—Ç–∏—Ö —Ñ—Ä–∞–Ω—Ü—É–∑—Å–∫–∏—Ö –±—É–ª–æ–∫';
      const result = languageDetectService.isRussian(text);

      expect(result.result).toEqual(true);
    });

    it('should return true if russian edge case', () => {
      const text =
        'üóÇ –ù–æ–º–µ—Ä —Ç–µ–ª–µ—Ñ–æ–Ω–∞\n' +
        '\n' +
        '–í–∞–º –Ω–µ–æ–±—Ö–æ–¥–∏–º–æ –ø–æ–¥—Ç–≤–µ—Ä–¥–∏—Ç—å –Ω–æ–º–µ—Ä —Ç–µ–ª–µ—Ñ–æ–Ω–∞ –¥–ª—è —Ç–æ–≥–æ, —á—Ç–æ–±—ã –∑–∞–≤–µ—Ä—à–∏—Ç—å –∏–¥–µ–Ω—Ç–∏—Ñ–∏–∫–∞—Ü–∏—é.\n' +
        '–î–ª—è —ç—Ç–æ–≥–æ –Ω–∞–∂–º–∏—Ç–µ –∫–Ω–æ–ø–∫—É –Ω–∏–∂–µ.';
      const result = languageDetectService.isRussian(text);

      expect(result.result).toEqual(true);
    });

    it('should return false if ukrainian edge case', () => {
      const text = '–ù–∞–∂–µ—Ä–ª–∏—Å—å –≥—Ä–µ—á–∫–∏ —ñ —Å–µ—Ä—É—Ç—å';
      const result = languageDetectService.isRussian(text);

      expect(result.result).toEqual(false);
    });

    it('should return false if ukrainian', () => {
      const text = '–Ü–Ω—Ç–µ—Ä–Ω–µ—Ç –±—É–ª–∏ –≤–∑—ñ—Ä—Ü–µ–º –¥–ª—è –Ω–∞—Å–ª—ñ–¥—É–≤–∞–Ω–Ω—è –¥–ª—è –≤—Å—ñ—Ö, —Ö—Ç–æ —Ö–æ—á–µ –≥–æ–≤–æ—Ä–∏—Ç–∏ —É–∫—Ä–∞—ó–Ω—Å—å–∫–æ—é –ø—Ä–∞–≤–∏–ª—å–Ω–æ';
      const result = languageDetectService.isRussian(text);

      expect(result.result).toEqual(false);
    });

    it('should except ukrainian for all messages', () => {
      const results = languageDetectUaMocks.map((text) => languageDetectService.detect(text));

      expect(results).toMatchSnapshot();
    });
  });

  describe('detect', () => {
    it('should detect russian', () => {
      const text = '—Å—ä–µ—à—å –µ—â–µ —ç—Ç–∏—Ö —Ñ—Ä–∞–Ω—Ü—É–∑—Å–∫–∏—Ö –±—É–ª–æ–∫';
      const result = languageDetectService.detect(text);

      expect(result).toMatchSnapshot();
    });

    it('should detect surzhik as ukrainian', () => {
      const text = "–î–æ–±—Ä–µ —É—Ç—Ä–æ. –Ø–∫ –¥—ñ–ª–∞? –æ–ø–ª–∞—Ç–∏—Ç–∏ –±—ñ–ª—î—Ç—ñ–∫–∏. –ø—Ä–µ–¥''—è–≤–∏—Ç–∏ —É–¥–æ—Å—Ç–æ–≤—ñ—Ä—î–Ω—ñ—î";
      const result = languageDetectService.detect(text);

      expect(result).toMatchSnapshot();
    });

    it('should detect ukrainian', () => {
      const text = '–Ü–Ω—Ç–µ—Ä–Ω–µ—Ç –±—É–ª–∏ –≤–∑—ñ—Ä—Ü–µ–º –¥–ª—è –Ω–∞—Å–ª—ñ–¥—É–≤–∞–Ω–Ω—è –¥–ª—è –≤—Å—ñ—Ö, —Ö—Ç–æ —Ö–æ—á–µ –≥–æ–≤–æ—Ä–∏—Ç–∏ —É–∫—Ä–∞—ó–Ω—Å—å–∫–æ—é –ø—Ä–∞–≤–∏–ª—å–Ω–æ';
      const result = languageDetectService.detect(text);

      expect(result).toMatchSnapshot();
    });

    it('should not detect russian when mixed', () => {
      const text =
        '—Å—ä–µ—à—å –µ—â–µ —ç—Ç–∏—Ö —Ñ—Ä–∞–Ω—Ü—É–∑—Å–∫–∏—Ö –±—É–ª–æ–∫. –Ü–Ω—Ç–µ—Ä–Ω–µ—Ç –±—É–ª–∏ –≤–∑—ñ—Ä—Ü–µ–º –¥–ª—è –Ω–∞—Å–ª—ñ–¥—É–≤–∞–Ω–Ω—è –¥–ª—è –≤—Å—ñ—Ö, —Ö—Ç–æ —Ö–æ—á–µ –≥–æ–≤–æ—Ä–∏—Ç–∏ —É–∫—Ä–∞—ó–Ω—Å—å–∫–æ—é –ø—Ä–∞–≤–∏–ª—å–Ω–æ.';
      const result = languageDetectService.detect(text);

      expect(result).toMatchSnapshot();
    });
  });
});
