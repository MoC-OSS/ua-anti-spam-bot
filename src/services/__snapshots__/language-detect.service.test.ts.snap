// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageDetectService detect should detect russian 1`] = `
[
  {
    "accuracy": 1,
    "lang": "ru",
  },
]
`;

exports[`LanguageDetectService detect should detect surzhik as ukrainian 1`] = `
[
  {
    "accuracy": 1,
    "lang": "uk",
  },
]
`;

exports[`LanguageDetectService detect should detect ukrainian 1`] = `
[
  {
    "accuracy": 1,
    "lang": "uk",
  },
]
`;

exports[`LanguageDetectService detect should not detect russian when mixed 1`] = `
[
  {
    "accuracy": 1,
    "lang": "uk",
  },
]
`;

exports[`LanguageDetectService isRussian should except ukrainian for all messages 1`] = `
[
  [
    {
      "accuracy": 0.18993,
      "lang": "ru",
    },
    {
      "accuracy": 0.18524000000000002,
      "lang": "uk",
    },
  ],
  [
    {
      "accuracy": 0.125,
      "lang": "ru",
    },
    {
      "accuracy": 0.1076,
      "lang": "uk",
    },
  ],
  [
    {
      "accuracy": 1,
      "lang": "uk",
    },
  ],
  [
    {
      "accuracy": 1,
      "lang": "uk",
    },
  ],
  [
    {
      "accuracy": 1,
      "lang": "uk",
    },
  ],
  [
    {
      "accuracy": 1,
      "lang": "uk",
    },
  ],
  [
    {
      "accuracy": 0.125,
      "lang": "uk",
    },
    {
      "accuracy": 0.10893749999999999,
      "lang": "ru",
    },
  ],
  [
    {
      "accuracy": 1,
      "lang": "uk",
    },
  ],
]
`;
